### PR TITLE
Restore skipped/failed in JSON callback, add custom facts

### LINKED ISSUE #37050: json callback plugin no longer contain information on some task statuses like 'skipped'
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

json callback plugin

##### ANSIBLE VERSION
```
ansible 2.4.2.0
  config file = /home/jackzhao/tmp/ansible-test/ansible.cfg
  configured module search path = [u'/home/jackzhao/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, Oct 11 2015, 17:47:16) [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)]
```

##### CONFIGURATION
using the `json` callback plugin

##### OS / ENVIRONMENT
RHEL7

##### SUMMARY
Prior to ansible 2.4, with the json module enabled, if a `shell` task gets skipped due to a conditional `when` statement, for instance, would result in a `skipped: True` key in the json output. After 2.4 (at least, as of 2.4.2.0) this field disappeared, and as a result from the json output alone there isn't a reliable way of determining whether a task was skipped.

I tracked this down to this change: https://github.com/ansible/ansible/commit/01b6c7c9c6b7459a3cb53ffc2fe02a8dcc1a3acc

... where the task queue manager now sends a 'clean copy' of the result object to the callback plugin where the 'clean copy' process removes the status keys. There's a comment there that says `# statuses are already reflected on the event type` which seems to be the reason for this change in behavior.

Except... if I'm just enabling the `json` plugin and reading the output of my ansible playbook command, I don't actually have access to the event type and thus this information is lost. If the purpose of the `json` plugin is to allow ansible play results to be read programmatically, then I imagine this behavior is not what we expect from the `json` plugin since whether a task was skipped is a pretty important piece of information regarding the run.

For now, I'm going to work around this by using a custom version of the json plugin that adds the event type to the dumped data based on which callback event happened. Just thought I bring it up here. Thanks!

### LINKED ISSUE #37184: JSON output plugin to display custom stats summary collected by set_stats module
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME
json callback output plugin

##### ANSIBLE VERSION

```
v2.0.0-0.5.beta3-19153-gd45ceb3a4f
```

##### OS / ENVIRONMENT
ANSIBLE_SHOW_CUSTOM_STATS=true env variable set or
show_custom_stats = True setting in ansible.cfg

##### SUMMARY
JSON output plugin is not displaying custom stats collected by set_stats module.

##### STEPS TO REPRODUCE

```yaml
---
- hosts: all
  connection: local
  gather_facts: no
  tasks:
    - name: Set a custom stat
      set_stats:
        data:
          my_stat: 4
```

##### EXPECTED RESULTS

In order to easily use custom stats, the summary of those should be displayed. The default output plugin shows the summary in a "CUSTOM STATS" section, like this:
```
$ ANSIBLE_STDOUT_CALLBACK=default ANSIBLE_SHOW_CUSTOM_STATS=true ansible-playbook -i inv pb.yml 

PLAY [all] *************************************************************************************************************************************************************************************************

TASK [Set a custom stat] ***********************************************************************************************************************************************************************************
ok: [test.example.com]

PLAY RECAP *************************************************************************************************************************************************************************************************
test.example.com           : ok=1    changed=0    unreachable=0    failed=0   


CUSTOM STATS: **********************************************************************************************************************************************************************************************

	RUN: { "my_stat": 4}

```
##### ACTUAL RESULTS

When using json output module, the custom stats are not summarized:

```
$ ANSIBLE_STDOUT_CALLBACK=json ANSIBLE_SHOW_CUSTOM_STATS=true ansible-playbook -i inv pb.yml 
{
    "plays": [
        {
            "play": {
                "id": "34e6d735-c355-074e-bdd0-000000000006",
                "name": "all"
            },
            "tasks": [
                {
                    "hosts": {
                        "test.example.com": {
                            "_ansible_no_log": false,
                            "ansible_stats": {
                                "aggregate": true,
                                "data": {
                                    "my_stat": 4
                                },
                                "per_host": false
                            },
                            "changed": false
                        }
                    },
                    "task": {
                        "id": "34e6d735-c355-074e-bdd0-000000000008",
                        "name": "Set a custom stat"
                    }
                }
            ]
        }
    ],
    "stats": {
        "test.example.com": {
            "changed": 0,
            "failures": 0,
            "ok": 1,
            "skipped": 0,
            "unreachable": 0
        }
    }
}
```

One could expect that when ANSIBLE_SHOW_CUSTOM_STATS is set:

- Per host summary of custom stats are present in the stats object, or alternatively, in custom_stats object to make a clear difference between custom and 'standard' stats.

- Per run stat summary is available via some other object (custom_run_stats perhaps?)

Like this with the example playbook:
```
  [...]
   "stats": {
        "test.example.com": {
            "changed": 0,
            "failures": 0,
            "ok": 1,
            "skipped": 0,
            "unreachable": 0
        }
    },
   "custom_stats": {
   },
   "custom_run_stats": {
       "my_stat": 4
   }
```

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY

This PR addresses 2 issues:

1. Restore skipped/failed keys. Fixes #37050
2. Display custom stats. Fixes #37184

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/plugins/callback/json.py

##### ANSIBLE VERSION

```
devel
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/plugins/callback/json.py [hunk 1]
```diff
@@ -15,10 +15,26 @@
     type: stdout
     requirements:
       - Set as stdout in config
+    options:
+      show_custom_stats:
+        name: Show custom stats
```
Comment (bcoca) [inline lib/ansible/plugins/callback/json.py]: bersion_abbed

Comment (sivel) [reply lib/ansible/plugins/callback/json.py]: added

**Comment by sivel** [comment]
@bcoca version_added as requested.

File: lib/ansible/plugins/callback/json.py [hunk 2]
```diff
@@ -72,13 +90,37 @@ def v2_playbook_on_stats(self, stats):
             s = stats.summarize(h)
             summary[h] = s
 
+        custom_stats = {}
+        if self._plugin_options.get('show_custom_stats', C.SHOW_CUSTOM_STATS) and stats.custom:  # fallback on constants for inherited plugins missing docs
```
Comment (bcoca) [inline lib/ansible/plugins/callback/json.py]: you don't need the constant, the config definition above already covers this

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/callback/json.py (hunk 2 above)
Comment (sivel) [reply lib/ansible/plugins/callback/json.py]: I had copied that directly from the default callback.  I've updated to use `self.get_option` instead.

Comment (bcoca) [reply lib/ansible/plugins/callback/json.py]: the default had to allow for '3rd party callback' subclassing and not using the new config framework

**Review by bcoca — APPROVED** [review]

**Pull request merged**